File: src/services/dependencyLens.test.ts
import { test } from "node:test";

import { expect } from "expect";

import { dependencyLenses, findDependencyEntries } from "./dependencyLens.ts";

const CONFIG = `{
  // Project configuration.
  "$schema": "./cappu.schema.json",
  "compilerOptions": {
    "outDir": "./build",
  },
  "packageSources": ["https://repo.maven.apache.org/maven2"],
  "dependencies": {
    "api": {
      "org.lib:core": "1.0",
    },
    "implementation": {
      // "org.commented:out": "9.9",
      "com.google.code.gson:gson": "2.10.1",
    },
  },
}
`;

test("dependency entries are found by their colon-bearing keys only", () => {
  const entries = findDependencyEntries(CONFIG);
  expect(entries.map(e => `${e.groupId}:${e.artifactId}@${e.version}`)).toEqual([
    "org.lib:core@1.0",
    "com.google.code.gson:gson@2.10.1",
  ]);
  // not matched: $schema/outDir keys (no colon in the key), the packageSources
  // url (an array element, not a key), the commented-out entry
  const gson = entries[1]!;
  expect(gson.line).toBe(13);
  expect(CONFIG.split("\n")[gson.line]!.slice(gson.startCharacter, gson.endCharacter)).toBe(
    '"com.google.code.gson:gson": "2.10.1"',
  );
});

test("a lens appears exactly for entries with a different newest version", async () => {
  const lookup = (groupId: string, artifactId: string): Promise<string | undefined> =>
    Promise.resolve(
      `${groupId}:${artifactId}` === "com.google.code.gson:gson" ? "2.14.0" : undefined,
    );
  const lenses = await dependencyLenses(CONFIG, lookup);
  expect(lenses).toHaveLength(1); // org.lib:core is unknown to the source: no lens
  expect(lenses[0]!.title).toBe("newer version: 2.14.0");
  expect(lenses[0]!.entry.artifactId).toBe("gson");
});

test("an up-to-date entry gets no lens", async () => {
  const lenses = await dependencyLenses(CONFIG, () => Promise.resolve("1.0"));
  expect(lenses.map(l => l.entry.artifactId)).toEqual(["gson"]); // 2.10.1 != 1.0
});

File: togo/internal/services/dependency_lens_test.go
package services

import (
	"strings"
	"testing"
)

// Port of src/services/dependencyLens.test.ts.

const depConfig = `{
  // Project configuration.
  "$schema": "./cappu.schema.json",
  "compilerOptions": {
    "outDir": "./build",
  },
  "packageSources": ["https://repo.maven.apache.org/maven2"],
  "dependencies": {
    "api": {
      "org.lib:core": "1.0",
    },
    "implementation": {
      // "org.commented:out": "9.9",
      "com.google.code.gson:gson": "2.10.1",
    },
  },
}
`

func TestFindDependencyEntries(t *testing.T) {
	entries := FindDependencyEntries(depConfig)
	var coords []string
	for _, e := range entries {
		coords = append(coords, e.GroupID+":"+e.ArtifactID+"@"+e.Version)
	}
	want := []string{"org.lib:core@1.0", "com.google.code.gson:gson@2.10.1"}
	if strings.Join(coords, ",") != strings.Join(want, ",") {
		t.Fatalf("entries = %v, want %v", coords, want)
	}
	gson := entries[1]
	if gson.Line != 13 {
		t.Errorf("gson line = %d, want 13", gson.Line)
	}
	lineText := strings.Split(depConfig, "\n")[gson.Line]
	if got := lineText[gson.StartChar:gson.EndChar]; got != `"com.google.code.gson:gson": "2.10.1"` {
		t.Errorf("span = %q", got)
	}
}

func TestDependencyLensNewerVersion(t *testing.T) {
	lookup := func(groupID, artifactID string) (string, bool) {
		if groupID+":"+artifactID == "com.google.code.gson:gson" {
			return "2.14.0", true
		}
		return "", false
	}
	lenses := DependencyLenses(depConfig, lookup)
	if len(lenses) != 1 || lenses[0].Title != "newer version: 2.14.0" || lenses[0].Entry.ArtifactID != "gson" {
		t.Errorf("lenses = %+v", lenses)
	}
}

func TestDependencyLensUpToDate(t *testing.T) {
	lenses := DependencyLenses(depConfig, func(_, _ string) (string, bool) { return "1.0", true })
	var arts []string
	for _, l := range lenses {
		arts = append(arts, l.Entry.ArtifactID)
	}
	if len(arts) != 1 || arts[0] != "gson" {
		t.Errorf("artifacts = %v, want [gson]", arts)
	}
}
